clock_control: stm32: mp13: add plls / mco & clock source selection

avolmat-st opened the pull request:
This PR adds support for PLL2 / PLL3 / PLL4 of the stm32mp13 as well as MCO1 and MCO2.

Peripheral clock selection code is as well added.

This work is done in order to allow usage of the DCMIPP (camera pipeline) PR #89407 / ST-MIPID02 PR #89764  on the stm32mp135f-dk board.

Commit:
drivers: clock: stm32-mco: support MCO with enable bit

On STM32MP13, a dedicated MCO enable bit within the MCO clock control
register must be set in order to activate the MCO.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
clock_control: stm32: add handling of clocks for the stm32mp13

Add enabled_clock, on / off and configure support for the clocks of
the stm32mp13. Describes the peripheral clock source selection.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
dts: arm: st: add MCOs and PLL2 to PLL4 in stm32mp13.dtsi

The stm32mp13 has 2 MCOs and 4 PLLs available.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
dts: bindings: stm32: add div-q and div-r on stm32mp13 plls

Depending on the PLL, all DIV-P / DIV-Q and DIV-R are available
on STM32MP13 PLLs.
Adjust valid range in order to be able to set for all 4 PLLs.
Clarify DT properties description.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

erwango commented on drivers/clock_control/clock_stm32_mco.c:
```diff
@@ -49,6 +49,11 @@ static int stm32_mco_init(const struct device *dev)
 		STM32_DT_CLKSEL_VAL_GET(pclken->enr) <<
 			STM32_DT_CLKSEL_SHIFT_GET(pclken->enr));
 
+#if defined(MCO_ENABLE_BIT)
```
Any way to use a Cube definition ?

avolmat-st replied:
Within cube, following 2 macros are available RCC_MCO1CFGR_MCO1ON and RCC_MCO2CFGR_MCO2ON.
The id of the MCO is part of the macro hence doesn't help to have generic code matching both of the MCOs.
I could replace the MCO_ENABLE_BIT by RCC_MCO1CFGR_MCO1ON, knowing that anyway if there is a MCO1ON bit then there will also be same enable bit on another MCO instance of the same soc.
RCC_MCO1CFGR_MCO1ON and RCC_MCO1CFGR_MCO1ON are both BIT(12).

At least this would avoid having to define a new MCO_ENABLE_BIT. What do you think ?

avolmat-st replied:
Turns out that such MCO enable bit are also necessary on the stm32f410 serie as well however in this latter case they are in same register but at different offset. So this method or Cube macro usage won't be possible as well.
I propose to keep it as is for the time being (MCO_ENABLE_BIT) and this will be reworked when MCO will have to be enabled in the stm32f410 context.

erwango commented on include/zephyr/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -195,6 +195,16 @@
 #define STM32_PLL_FRACN_VALUE	DT_PROP_OR(DT_NODELABEL(pll), fracn, 1)
 #endif
 
+#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(pll), st_stm32mp13_pll_clock, okay)
+#define STM32_PLL_ENABLED	1
+#define STM32_PLL_M_DIVISOR	DT_PROP(DT_NODELABEL(pll), div_m)
+#define STM32_PLL_N_MULTIPLIER	DT_PROP(DT_NODELABEL(pll), mul_n)
+#define STM32_PLL_P_ENABLED	DT_NODE_HAS_PROP(DT_NODELABEL(pll), div_p)
+#define STM32_PLL_P_DIVISOR	DT_PROP_OR(DT_NODELABEL(pll), div_p, 1)
+#define STM32_PLL_FRACV_ENABLED	DT_NODE_HAS_PROP(DT_NODELABEL(pll), frac_v)
+#define STM32_PLL_FRACV_VALUE	DT_PROP_OR(DT_NODELABEL(pll), frac_v, 1)
+#endif
```
Please reuse the above section to avoid duplication.
(FYI, not all the properties above are available for all the compats, if that was you concern. Errors are cached at other levels).

avolmat-st replied:
indeed, that was the reason for this duplication. Ok, merged with existing onces for PLL1 / PLL2 and PLL3. Kept it different for PLL4 which is only available for N6 with different bindings.

erwango commented on include/zephyr/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -195,6 +195,16 @@
+#define STM32_PLL_N_MULTIPLIER	DT_PROP(DT_NODELABEL(pll), mul_n)
+#define STM32_PLL_P_ENABLED	DT_NODE_HAS_PROP(DT_NODELABEL(pll), div_p)
+#define STM32_PLL_P_DIVISOR	DT_PROP_OR(DT_NODELABEL(pll), div_p, 1)
+#define STM32_PLL_FRACV_ENABLED	DT_NODE_HAS_PROP(DT_NODELABEL(pll), frac_v)
```
I'm slightly puzzled by the lack of consistency between `frac-v` and already available `fracn`.
I agree that `fracn` is the intruder here. We'll likely need to change in a follow up PR for consistency.

Apply everywhere unless there are strong reasons to go against.

avolmat-st replied:
Done, I added a commit to exchange FRACV into fracn on the mp13. At some places within comment I kept the name FRACV since this is the name of the register field.

Commit:
drivers: clock: stm32mp13: rename frac-v binding into fracn

Rename the frac-v PLL binding into fracn in order to make it
consistent with other STM32 PLL bindings.
This commit also correct the range which should be 0 - 8191.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

etienne-lms commented on dts/bindings/clock/st,stm32mp13-pll-clock.yaml:
```diff
@@ -80,8 +80,8 @@ properties:
         Valid range: 1 - 128
 
-  frac-v:
+  fracn:
```
Indeed few other STM32 SoCs use `fracn` as DT property, but their related SoC reference manual refer to a **FRACN** value whereas STM32MP13 SoC reference manual refer to **FRACV** value. I think `frac-v` as DT property better reflect the manual. But that is not a strong opinion.

arnopo replied:
I suggest to specify in description that it coresponds to FRACV as down for PLL in previous commit

avolmat-st replied:
The description already state the following:

-  frac-v:
+  fracn:
     type: int
     description: |
         PLLx FRACV fractional latch
-        Valid range: 1 - 8192
+        Valid range: 0 - 8191

etienne-lms replied:
Ok, fine with me.

etienne-lms reviewed (commented):
LGTM with a minor question.

arnopo commented on dts/bindings/clock/st,stm32mp13-pll-clock.yaml:
```diff
@@ -28,6 +28,13 @@ description: |
   (refer to the stm32mp13 reference manual for details)
 
+  div-m, div-p, div-q, div-r valid ranges are identicals for all 4 PLLs
+  mul-n valid range depends on the PLL
```
I would use here the Name defined in stm32mp13 reference manual but also in formula above
```
DIVM1, DIVP, DIVQ and DIVR have identical valid ranges for all 4 PLLs.
DIVN valid range depends on the PLL
```
or use aka as below
```
div-m (aka DIVM1) , div-p (aka DIVP), div-q (aka DIVQ), div-r (aka DIVR)  have identical valid ranges for all 4 PLLs.
mul-n (aka DIVN) valid range depends on the PLL

```

arnopo commented on dts/bindings/clock/st,stm32mp13-pll-clock.yaml:
```diff
@@ -53,14 +60,26 @@ properties:
     description: |
         PLLx multiplication factor for VCO
-        Valid range: 31 - 125
+        Valid range: 25 - 200
```
specify that it is DIVN
`PLLx DIVN multiplication factor for VCO`

arnopo reviewed (commented):
minor comments else LGTM

etienne-lms commented on dts/bindings/clock/st,stm32mp13-pll-clock.yaml:
```diff
@@ -44,23 +52,34 @@ properties:
     description: |
-        Prescaler for PLLx
-        input clock
+        PLLx DIVM1 division factor of the input clock divider
```
Actually, the value of the property is the value applied to PLLx DIVM1 plus 1.
So either not mention `DIVM1` here (and ditto for `DIVN`, ... below) or this needs some rewording, something like:
```
        PLLx division factor of the input clock divider (aka DIVM1 + 1)
